File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.java
class Solution {
    public int pairSum(ListNode head) {

        // STEP 1: Find middle of linked list
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;        // move 1 step
            fast = fast.next.next;   // move 2 steps
        }

        /*
        Example: [5,4,2,1]

        head → 5 -> 4 -> 2 -> 1

        After loop:
        slow → 2
        (middle node)

        First half : 5 -> 4
        Second half: 2 -> 1
        */


        // STEP 2: Reverse second half
        ListNode prev = null;

        while (slow != null) {

            // save next node
            ListNode nextNode = slow.next;

            /*
            First iteration:
            slow → 2 -> 1
            nextNode → 1
            */

            // reverse link
            slow.next = prev;

            /*
            2 -> null
            because prev = null initially
            */

            // move prev forward
            prev = slow;

            /*
            prev → 2 -> null
            */

            // move slow forward
            slow = nextNode;

            /*
            slow → 1
            */


            /*
            Second iteration:

            nextNode = null

            reverse:
            1 -> 2 -> null

            prev → 1 -> 2 -> null

            slow → null

            loop ends
            */
        }

        /*
        After reversing:

        prev → 1 -> 2 -> null

        IMPORTANT:
        prev becomes NEW HEAD
        of reversed second half
        */


        // STEP 3: Compare twin nodes
        int maxSum = 0;

        ListNode first = head;
        ListNode second = prev;

        /*
        first → 5 -> 4 -> 2
        second → 1 -> 2

        Why?

        first = original first half
        second = reversed second half

        So twins align automatically:

        5 ↔ 1
        4 ↔ 2
        */

        while (second != null) {

            maxSum = Math.max(
                maxSum,
                first.val + second.val
            );

            /*
            Iteration 1:
            5 + 1 = 6

            maxSum = 6
            */

            first = first.next;
            second = second.next;

            /*
            first → 4
            second → 2
            */

            /*
            Iteration 2:
            4 + 2 = 6

            maxSum = 6
            */
        }

        return maxSum;
    }
}

File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:

        # -----------------------------------
        # STEP 1: Find middle of linked list
        # -----------------------------------

        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next       # move 1 step
            fast = fast.next.next  # move 2 steps

        """
        Example:
        head = [5,4,2,1]

        Linked List:
        5 -> 4 -> 2 -> 1

        After loop:

        slow → 2
        fast → None

        So:
        First half  = 5 -> 4
        Second half = 2 -> 1
        """



        # -----------------------------------
        # STEP 2: Reverse second half
        # -----------------------------------

        prev = None

        while slow:

            # save next node
            next_node = slow.next

            """
            First iteration:

            slow → 2 -> 1
            next_node → 1
            """

            # reverse link
            slow.next = prev

            """
            Before:
            2 -> 1

            After:
            2 -> None
            because prev = None initially
            """

            # move prev forward
            prev = slow

            """
            prev → 2 -> None
            """

            # move slow forward
            slow = next_node

            """
            slow → 1
            """

            """
            Second iteration:

            next_node = None

            reverse:
            1 -> 2 -> None

            prev → 1 -> 2 -> None

            slow → None
            loop stops
            """

        """
        Final reversed second half:

        prev → 1 -> 2 -> None

        IMPORTANT:
        prev is now the HEAD
        of reversed linked list
        """



        # -----------------------------------
        # STEP 3: Compare twin sums
        # -----------------------------------

        max_sum = 0

        first = head
        second = prev

        """
        first  → 5 -> 4 -> 2
        second → 1 -> 2

        Why?

        second points to reversed half.

        Twin pairs become aligned:

        5 ↔ 1
        4 ↔ 2
        """

        while second:

            # calculate twin sum
            twin_sum = first.val + second.val

            # update maximum
            max_sum = max(max_sum, twin_sum)

            """
            Iteration 1:
            5 + 1 = 6

            max_sum = 6
            """

            # move both pointers
            first = first.next
            second = second.next

            """
            first → 4
            second → 2

            Iteration 2:
            4 + 2 = 6

            max_sum = 6
            """

        return max_sum